metrics/api/src/main/java/io/helidon/metrics/api/AbstractRegistry.java

```diff
@@ -0,0 +1,808 @@
+        return getOrRegisterMetric(metadata, Histogram.class, tags);
+    }
+
+    @Override
+    public Meter meter(String name) {
+        return meter(name, NO_TAGS);
+    }
+
+    @Override
+    public Meter meter(Metadata metadata) {
+        return meter(metadata, NO_TAGS);
+    }
+
+    @Override
+    public Meter meter(String name, Tag... tags) {
+        return getOrRegisterMetric(name, Meter.class, tags);
+    }
+
+    @Override
+    public Meter meter(Metadata metadata, Tag... tags) {
+        return getOrRegisterMetric(metadata, Meter.class, tags);
+    }
+
+    @Override
+    public Timer timer(String name) {
+        return timer(name, NO_TAGS);
+    }
+
+    @Override
+    public Timer timer(Metadata metadata) {
+        return timer(metadata, NO_TAGS);
+    }
+
+    @Override
+    public Timer timer(String name, Tag... tags) {
+        return getOrRegisterMetric(name, Timer.class, tags);
+    }
+
+    @Override
+    public Timer timer(Metadata metadata, Tag... tags) {
+        return getOrRegisterMetric(metadata, Timer.class, tags);
+    }
+
+    @Override
+    public ConcurrentGauge concurrentGauge(String name) {
+        return concurrentGauge(name, NO_TAGS);
+    }
+
+    @Override
+    public ConcurrentGauge concurrentGauge(Metadata metadata) {
+        return concurrentGauge(metadata, NO_TAGS);
+    }
+
+    @Override
+    public ConcurrentGauge concurrentGauge(String name, Tag... tags) {
+        return getOrRegisterMetric(name, ConcurrentGauge.class, tags);
+    }
+
+    @Override
+    public ConcurrentGauge concurrentGauge(Metadata metadata, Tag... tags) {
+        return getOrRegisterMetric(metadata, ConcurrentGauge.class, tags);
+    }
+
+    @Override
+    public SimpleTimer simpleTimer(String name) {
+        return simpleTimer(name, NO_TAGS);
+    }
+
+    @Override
+    public SimpleTimer simpleTimer(Metadata metadata) {
+        return simpleTimer(metadata, NO_TAGS);
+    }
+
+    @Override
+    public SimpleTimer simpleTimer(String name, Tag... tags) {
+        return getOrRegisterMetric(name, SimpleTimer.class, tags);
+    }
+
+    @Override
+    public SimpleTimer simpleTimer(Metadata metadata, Tag... tags) {
+        return getOrRegisterMetric(metadata, SimpleTimer.class, tags);
+    }
+
+    /**
+     * Removes a metric by name. Synchronized for atomic update of more than one internal map.
+     *
+     * @param name Name of the metric.
+     * @return Outcome of removal.
+     */
+    @Override
+    public synchronized boolean remove(String name) {
+        final List<Map.Entry<MetricID, M>> doomedMetrics = getMetricsByName(name);
+        if (doomedMetrics.isEmpty()) {
+            return false;
+        }
+
+        final boolean result = doomedMetrics.stream()
+                .peek(entry -> entry.getValue().markAsDeleted())
+                .map(entry -> allMetrics.remove(entry.getKey()) != null)
+                .reduce((a, b) -> a || b)
+                .orElse(false);
```

Comment:
Replacing with
```
        boolean result = false;
        for (Map.Entry<MetricID, M> doomedMetric : doomedMetrics) {
            doomedMetric.getValue().markAsDeleted();
            result |= allMetrics.remove(doomedMetric.getKey()) != null;
        }
```
Perhaps things here could have been handled a little more cleanly overall, but as it stands the method `getMetricsByName` which supplies the entries in `doomedMetrics` returns a `List`, the `Map.Entry<MetricID, M>` elements of which are not tied to any original map. They are just convenient pairings of a metric ID with its metric which happens to use `Map.Entry` to put them together rather than some private static class with two fields. So `Iterator.remove` is not an option here, at least not as currently implemented.

Reply:
Oh, OK